### PR TITLE
Send state message when components are removed

### PR DESCRIPTION
**Pull request by nkoenig**
Signed-off-by: Nate Koenig <nate@openrobotics.org>

# 🦟 Bug fix

## Summary

The `WorldCmdPose` was being transmitted to the GUI and back to the Server, resulting in bad behavior. See [here](https://github.com/ignitionrobotics/ign-gazebo/pull/1231#issuecomment-985923720).

This PR changes the SceneBroadcaster to send a new state message when components are removed.

Before I write tests, I'd like to get some feedback on this approach.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Updated documentation (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Review by chapulina — COMMENTED** [review]
Nice! We just need a unit test.

**Comment by nkoenig** [comment]
> Nice! We just need a unit test.

Just checking that you don't think there are unforeseen side effects of this?

**Comment by chapulina** [comment]
> Just checking that you don't think there are unforeseen side effects of this?

I can't think of any unwanted effects, the worst I can think of is that we're publishing more state messages, but that should be negligible if there aren't components being removed every iteration.

**Comment by azeey** [comment]
We have been preferring to zero out components instead of removing them for performance reasons, so I don't think we have any components that get removed every iteration. So I think this approach might work as a work around. My proposal for a more long term solution is to have a scheme where components are requested for removal, just like entities, and their removal happens after all the `*Update` phases have completed. I think this would remove the need to keep `WorldPoseCmd` around for an extra iteration.

**Comment by nkoenig** [comment]
Tests have been added in b2ad027

File: test/integration/scene_broadcaster_system.cc [hunk 1]
```diff
@@ -612,6 +616,95 @@ TEST_P(SceneBroadcasterTest, StateStatic)
...
+                  ignition::gazebo::components::Pose>(
+          [&](const ignition::gazebo::Entity &_entity,
+              const ignition::gazebo::components::Model *,
+              const ignition::gazebo::components::Name *_name,
+              const ignition::gazebo::components::Pose *)->bool
+          {
+            if (_name->Data() == "box")
+            {
+              _ecm.RemoveComponent<ignition::gazebo::components::Pose>(_entity);
+              }
```
Comment (chapulina) [inline test/integration/scene_broadcaster_system.cc:651]:
```suggestion
            }
```

File: test/integration/scene_broadcaster_system.cc [hunk 2]
```diff
@@ -612,6 +616,95 @@ TEST_P(SceneBroadcasterTest, StateStatic)
...
+  sleep = 0u;
+  received = false;
+  hasState = false;
+  server.RunOnce(true);
+  // cppcheck-suppress unmatchedSuppression
+  // cppcheck-suppress knownConditionTrueFalse
+  while (!received && sleep++ < maxSleep)
+    IGN_SLEEP_MS(100);
+  EXPECT_TRUE(received);
+  EXPECT_TRUE(hasState);
```
Comment (chapulina) [inline test/integration/scene_broadcaster_system.cc:705]: We could check specifically that the message contains a removed component

**Review by chapulina — APPROVED** [review]

File: test/integration/scene_broadcaster_system.cc (hunk 2 above)
Comment (nkoenig) [reply test/integration/scene_broadcaster_system.cc:737]: Done in 326a841

File: test/integration/scene_broadcaster_system.cc (hunk 1 above)
Comment (nkoenig) [reply test/integration/scene_broadcaster_system.cc]: Done in 326a841

**Review by chapulina — APPROVED** [review]

**Pull request merged**